resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -276,15 +276,48 @@ private static String hostnameWithDot(String name) {
         return name + '.';
     }
 
-    private void internalResolve(String name, Promise<List<T>> promise) {
-        for (;;) {
-            // Resolve from cnameCache() until there is no more cname entry cached.
-            String mapping = cnameCache().get(hostnameWithDot(name));
-            if (mapping == null) {
+    // Resolve the final name from the CNAME cache until there is nothing to follow anymore. This also
+    // guards against loops in the cache but early return once a loop is detected.
+    private String cnameResolveFromCache(String name) {
+        DnsCnameCache cnameCache = cnameCache();
+        String first = cnameCache.get(hostnameWithDot(name));
+        if (first == null) {
+            // Nothing in the cache at all
+            return name;
+        }
+
+        String mapping = cnameCache.get(hostnameWithDot(first));
+        if (mapping == null) {
+            // Nothing else to follow, return first match.
+            return first;
+        }
+
+        if (first.equals(mapping)) {
+            // Loop detected.... early return.
+            return first;
+        }
+
+        // Detect loops using a HashSet. We use this as last resort implementation to reduce allocations in the most
```

Comment:
Just code style, so not necessary: I would pack this "uncommon case" in a separate method

Reply:
@franz1981 done... PTAL again